core/server/worker/src/main/java/alluxio/worker/netty/DataServerUfsBlockReadHandler.java

```diff
@@ -0,0 +1,130 @@
+
+import com.codahale.metrics.Counter;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * This handler handles UFS block read request. Check more information in
+ * {@link DataServerReadHandler}.
+ */
+@NotThreadSafe
+public final class DataServerUfsBlockReadHandler extends DataServerReadHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(DataServerUfsBlockReadHandler.class);
+
+  /** The Block Worker. */
+  private final BlockWorker mWorker;
+
+  /**
+   * The block read request internal representation.
+   */
+  private final class UfsBlockReadRequestInternal extends ReadRequestInternal {
+    public BlockReader mBlockReader = null;
+
+    /**
+     * Creates an instance of {@link UfsBlockReadRequestInternal}.
+     *
+     * @param request the block read request
+     * @throws Exception if it fails to create the object
+     */
+    public UfsBlockReadRequestInternal(Protocol.ReadRequest request) throws Exception {
+      mId = request.getId();
+      mStart = request.getOffset();
+      mEnd = mStart + request.getLength();
+      mBlockReader =
+          mWorker.readUfsBlock(request.getSessionId(), mId, mStart, request.getNoCache());
+      // Note that we do not need to seek to offset since the block worker is created at the offset.
+    }
+
+    @Override
+    public void close() {
+      if (mBlockReader != null)  {
+        try {
+          mBlockReader.close();
+        } catch (Exception e) {
+          LOG.warn("Failed to close block reader for block {}.", mId);
+        }
+      }
+    }
+  }
+
+  /**
+   * Creates an instance of {@link DataServerReadHandler}.
+   *
+   * @param executorService the executor service to run {@link PacketReader}s
+   * @param blockWorker the block worker
+   */
+  public DataServerUfsBlockReadHandler(ExecutorService executorService, BlockWorker blockWorker) {
+    super(executorService);
+    mWorker = blockWorker;
+  }
+
+  @Override
+  protected boolean acceptMessage(Object object) {
+    if (!super.acceptMessage(object)) {
+      return false;
+    }
+    Protocol.ReadRequest request = ((RPCProtoMessage) object).getMessage().getMessage();
+    return request.getType() == Protocol.RequestType.UFS_BLOCK;
+  }
+
+  @Override
+  protected void initializeRequest(Protocol.ReadRequest request) throws Exception {
+    mRequest = new UfsBlockReadRequestInternal(request);
+  }
+
+  @Override
+  protected DataBuffer getDataBuffer(Channel channel, long offset, int len) throws IOException {
+    BlockReader blockReader = ((UfsBlockReadRequestInternal) mRequest).mBlockReader;
+    ByteBuf buf = channel.alloc().buffer(len, len);
+    try {
+      while (buf.writableBytes() > 0 && blockReader.transferTo(buf) != -1) {
+      }
+      return new DataNettyBufferV2(buf);
```

Comment:
@peisun1115 could you document here that `buf` will be released by netty buffer?